doc/packet_trimming/packet-trimming-design.md

```diff
@@ -0,0 +1,783 @@
+```abnf
+; defines schema for switch trimming configuration capabilities
+key = SWITCH_CAPABILITY|switch ; must be unique
+
+; field                                      = value
+SWITCH_TRIMMING_CAPABLE                      = capability-knob ; specifies whether switch is trimming capable
+SWITCH|PACKET_TRIMMING_QUEUE_RESOLUTION_MODE = queue-mode-list ; queue mapping mode capabilities
+
+; value annotations
+capability-knob = "true" / "false"
+queue-mode      = "STATIC" / "DYNAMIC"
+queue-mode-list = "" / "N/A" / queue-mode [ 1*( "," queue-mode ) ]
+```
+
+### 2.4.3 Data sample
+
+**Config DB:**
+```bash
+redis-cli -n 4 HGETALL 'SWITCH_TRIMMING|GLOBAL'
+1) "size"
+2) "128"
+3) "dscp_value"
+4) "48"
+5) "queue_index"
+6) "6"
+
+redis-cli -n 4 HGETALL 'BUFFER_PROFILE|q_lossy_trim_profile'
+1) "packet_discard_action"
+2) "trim"
+
+redis-cli -n 4 HGETALL 'ACL_RULE|TRIM_TABLE|TRIM_RULE'
+1) "PACKET_ACTION"
+2) "DISABLE_TRIM"
+```
+
+**State DB:**
+```bash
+redis-cli -n 6 HGETALL 'SWITCH_CAPABILITY|switch'
+ 1) "SWITCH_TRIMMING_CAPABLE"
+ 2) "true"
+ 3) "SWITCH|PACKET_TRIMMING_QUEUE_RESOLUTION_MODE"
+ 4) "STATIC,DYNAMIC"
+```
+
+### 2.4.4 Configuration sample
+
+**Packet trimming and filtering:**
+```json
+{
+    "ACL_TABLE_TYPE": {
+        "TRIMMING_L3": {
+            "MATCHES": [
+                "SRC_IP"
+            ],
+            "ACTIONS": [
+                "DISABLE_TRIM_ACTION"
+            ],
+            "BIND_POINTS": [
+                "PORT"
+            ]
+        }
+    },
+    "ACL_TABLE": {
+        "TRIM_TABLE": {
+            "POLICY_DESC": "Packet trimming",
+            "TYPE": "TRIMMING_L3",
+            "STAGE": "INGRESS",
+            "PORTS": [
+                "Ethernet0"
+            ]
+        }
+    },
+    "ACL_RULE": {
+        "TRIM_TABLE|TRIM_RULE": {
+            "PRIORITY": "999",
+            "SRC_IP": "1.1.1.1/32",
+            "PACKET_ACTION": "DISABLE_TRIM"
+        }
+    },
+    "BUFFER_POOL": {
+        "egress_lossy_pool": {
+            "mode": "dynamic",
+            "type": "egress"
+        }
+    },
+    "BUFFER_PROFILE": {
+        "q_lossy_trim_profile": {
+            "dynamic_th": "3",
+            "pool": "egress_lossy_pool",
+            "size": "0",
+            "packet_discard_action": "trim"
+        }
+    },
+    "BUFFER_QUEUE": {
+        "Ethernet0|3": {
+            "profile": "q_lossy_trim_profile"
+        }
+    },
+    "SWITCH_TRIMMING": {
+        "GLOBAL": {
+            "size": "128",
+            "dscp_value": "48",
+            "queue_index": "6"
+        }
+    }
+}
+```
+
+### 2.4.5 Initial configuration
+
+No special handling is required: disabled by default
+
+### 2.4.6 Configuration migration
+
+No special handling is required: adding new and extending the existing schema
+
+## 2.5 Flows
+
+### 2.5.1 Config section
+
+### 2.5.1.1 Switch trimming update
+
+![Switch trimming update flow](images/pt_switch_flow.svg "Figure 3: Switch trimming update flow")
+
```

Comment:
Please bring out the fact that a SONiC CLI operation will check the STATE\_DB before pushing to CONFIG\_DB.

Reply:
@rck-innovium fixed